### PR TITLE
Generalize transport benchmark

### PR DESCRIPTION
**Pull request by TaWeiTu**
Generalize bm_chttp2_transport.cc so that binder transport can be benchmarked using the same framework. This is done by extracting common benchmark logics out to a separate file `bm_transport.h` and make all functions generic over a Fixture. We can then implement a Fixture class for binder to make binder transport benchmark works (in upcoming PR).

### CODE REVIEW

**Comment by TaWeiTu** [comment]
Microbenchmarks Diff CI is failing due to the change in benchmark name. WIll try to set up an internal CL as well later.

File: test/cpp/microbenchmarks/bm_transport.cc [hunk 1]
```diff
@@ -591,9 +742,13 @@ static void BM_TransportStreamRecv(benchmark::State& state) {
   std::unique_ptr<TestClosure> c =
       MakeTestClosure([&](grpc_error_handle /*error*/) {
         if (!state.KeepRunning()) return;
-        // force outgoing window to be yuge
-        s->chttp2_stream()->flow_control->TestOnlyForceHugeWindow();
-        f.chttp2_transport()->flow_control->TestOnlyForceHugeWindow();
+        if (std::is_same<Fixture, Chttp2Fixture>::value) {
+          // force outgoing window to be yuge
```
Comment (sifmelcara) [inline test/cpp/microbenchmarks/bm_transport.cc:746]: 'yuge' misspelled?

File: test/cpp/microbenchmarks/bm_transport.cc [hunk 2]
```diff
@@ -680,7 +835,8 @@ static void BM_TransportStreamRecv(benchmark::State& state) {
   track_counters.Finish(state);
   grpc_slice_unref(incoming_data);
 }
-BENCHMARK(BM_TransportStreamRecv)->Range(0, 128 * 1024 * 1024);
+BENCHMARK_TEMPLATE(BM_TransportStreamRecv, Chttp2Fixture)
+    ->Range(0, 128 * 1024 * 1024);
```
Comment (sifmelcara) [inline test/cpp/microbenchmarks/bm_transport.cc:840]: No benchmark of binder transport here?

**Review by sifmelcara — COMMENTED** [review]
I would suggest to split the changes into multiple files (move the code shared between chttp2 and binder transport benchmark to a new file, and have benchmark of chttp2 and binder transport in different `.cc` file)

Also we probably want to split this into 2 PRs. The first PR do the generalization/refactoring and the second PR add binder transport benchmark

File: test/cpp/microbenchmarks/bm_transport.cc (hunk 1 above)
Comment (TaWeiTu) [reply test/cpp/microbenchmarks/bm_transport.cc]: Fixed

**Comment by TaWeiTu** [comment]
> I would suggest to split the changes into multiple files (move the code shared between chttp2 and binder transport benchmark to a new file, and have benchmark of chttp2 and binder transport in different `.cc` file)
> 
> Also we probably want to split this into 2 PRs. The first PR do the generalization/refactoring and the second PR add binder transport benchmark

Sure. Now this PR only contains the generalization part. Binder transport benchmark will be added later.

**Comment by ctiller** [comment]
I left a comment on the wrong PR! But, I'd suggest forking the code and not trying to make a general framework here.

**Comment by TaWeiTu** [comment]
> I left a comment on the wrong PR! But, I'd suggest forking the code and not trying to make a general framework here.

Sure, if you are fine with duplicated code here, then I think it'd be easier to just fork the code. I'll abandon this PR once #27085 is updated.

**Comment by ctiller** [comment]
Yeah, I think overall maintenance gets easier with a little duplication
here.

**Pull request closed without merging**